File: App/CallSounds.swift
import AVFoundation
import UIKit

/// Connect / DM UI audio — ringback, selectable ringtones & notification tones (full ambient pack).
/// Honors Settings → Sounds master toggle and iOS Silent switch (`.ambient`).
@MainActor
enum CallSounds {
    enum Ringtone: String, CaseIterable, Identifiable {
        case `default` = "ringtone"
        case gold = "ringtone_gold"
        case pulse = "ringtone_pulse"
        case chime = "ringtone_chime"
        case ambientAurora = "ringtone_ambient_aurora"
        case ambientBloom = "ringtone_ambient_bloom"
        case ambientGlass = "ringtone_ambient_glass"
        case ambientNight = "ringtone_ambient_night"
        case ambientRain = "ringtone_ambient_rain"
        case ambientSpace = "ringtone_ambient_space"

        var id: String { rawValue }
        var title: String {
            switch self {
            case .default: return "Default"
            case .gold: return "Gold"
            case .pulse: return "Pulse"
            case .chime: return "Chime"
            case .ambientAurora: return "Ambient · Aurora"
            case .ambientBloom: return "Ambient · Bloom"
            case .ambientGlass: return "Ambient · Glass"
            case .ambientNight: return "Ambient · Night"
            case .ambientRain: return "Ambient · Rain"
            case .ambientSpace: return "Ambient · Space"
            }
        }
    }

    enum NotificationTone: String, CaseIterable, Identifiable {
        case message = "notif_message"
        case call = "notif_call"
        case delivered = "notif_delivered"
        case read = "notif_read"
        case ambientSoft = "notif_ambient_soft"
        case ambientGlow = "notif_ambient_glow"
        case ambientDrop = "notif_ambient_drop"
        case ambientPetal = "notif_ambient_petal"
        case ambientWind = "notif_ambient_wind"

        var id: String { rawValue }
        var title: String {
            switch self {
            case .message: return "Message"
            case .call: return "Call"
            case .delivered: return "Delivered"
            case .read: return "Read"
            case .ambientSoft: return "Ambient · Soft"
            case .ambientGlow: return "Ambient · Glow"
            case .ambientDrop: return "Ambient · Drop"
            case .ambientPetal: return "Ambient · Petal"
            case .ambientWind: return "Ambient · Wind"
            }
        }
    }

    private static var oneShot: AVAudioPlayer?
    private static var loopPlayer: AVAudioPlayer?

    static var playUISounds: Bool {
        get { UserDefaults.standard.object(forKey: "playUISounds") as? Bool ?? true }
        set { UserDefaults.standard.set(newValue, forKey: "playUISounds") }
    }

    static var selectedRingtone: Ringtone {
        get { Ringtone(rawValue: UserDefaults.standard.string(forKey: "selectedRingtone") ?? "") ?? .default }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: "selectedRingtone") }
    }

    static var selectedNotificationTone: NotificationTone {
        get { NotificationTone(rawValue: UserDefaults.standard.string(forKey: "selectedNotificationTone") ?? "") ?? .message }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: "selectedNotificationTone") }
    }

    private static func prepareSession() {
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.ambient, mode: .default, options: [.mixWithOthers, .duckOthers])
        try? session.setActive(true)
    }

    private static func url(named name: String) -> URL? {
        Bundle.main.url(forResource: name, withExtension: "mp3", subdirectory: "Sounds")
            ?? Bundle.main.url(forResource: name, withExtension: "mp3")
            ?? Bundle.main.url(forResource: name, withExtension: "mp3", subdirectory: "Resources/Sounds")
    }

    private static func playOneShot(named name: String, volume: Float = 0.7) {
        guard playUISounds else { return }
        guard UIApplication.shared.applicationState != .background else { return }
        prepareSession()
        guard let url = url(named: name) else { return }
        do {
            oneShot = try AVAudioPlayer(contentsOf: url)
            oneShot?.volume = min(max(volume, 0.05), 0.85)
            oneShot?.prepareToPlay()
            oneShot?.play()
        } catch {}
    }

    private static func playLoop(named name: String, volume: Float = 0.45) {
        guard playUISounds else { return }
        prepareSession()
        stopLoop()
        guard let url = url(named: name) else { return }
        do {
            loopPlayer = try AVAudioPlayer(contentsOf: url)
            loopPlayer?.numberOfLoops = -1
            loopPlayer?.volume = min(max(volume, 0.05), 0.55)
            loopPlayer?.prepareToPlay()
            loopPlayer?.play()
        } catch {}
    }

    static func stopLoop() {
        loopPlayer?.stop()
        loopPlayer = nil
    }

    static func startRingback() { playLoop(named: "ringback_outgoing", volume: 0.42) }
    static func stopRingback() { stopLoop() }

    static func playRingtone(preview: Bool = false) {
        if preview {
            stopLoop()
            playOneShot(named: selectedRingtone.rawValue, volume: 0.65)
        } else {
            playLoop(named: selectedRingtone.rawValue, volume: 0.5)
        }
    }

    static func callInitiated() {
        startRingback()
        playOneShot(named: "call_initiated", volume: 0.55)
        CallHaptics.outgoingStart()
        CallHaptics.startRingingPulse()
    }

    static func callConnecting() { CallHaptics.connecting() }

    static func callConnected() {
        stopRingback()
        stopLoop()
        CallHaptics.stopRingingPulse()
        playOneShot(named: "call_connected", volume: 0.55)
        CallHaptics.connected()
    }

    static func hangup() {
        stopRingback()
        stopLoop()
        CallHaptics.stopRingingPulse()
        playOneShot(named: "hangup", volume: 0.65)
        CallHaptics.disconnected()
    }

    static func busyOrFailed() {
        stopRingback()
        stopLoop()
        CallHaptics.stopRingingPulse()
        playOneShot(named: "busy_tone", volume: 0.55)
        CallHaptics.busyOrFailed()
    }

    static func messageSent() {
        playOneShot(named: "message_sent", volume: 0.55)
        CallHaptics.messageSent()
    }

    static func attachmentSent() {
        playOneShot(named: "attachment_sent", volume: 0.55)
        CallHaptics.attachmentSent()
    }

    static func voiceNoteSent() {
        playOneShot(named: "voicenote_sent", volume: 0.6)
        CallHaptics.voiceNoteSent()
    }

    static func typingPing() { playOneShot(named: "typing_ping", volume: 0.3) }

    static func inboundNotification() {
        playOneShot(named: selectedNotificationTone.rawValue, volume: 0.6)
        CallHaptics.inboundMessage()
    }

    static func delivered() {
        playOneShot(named: "notif_delivered", volume: 0.4)
        CallHaptics.delivered()
    }

    static func read() {
        playOneShot(named: "notif_read", volume: 0.35)
        CallHaptics.read()
    }

    static func previewNotification(_ tone: NotificationTone) {
        playOneShot(named: tone.rawValue, volume: 0.6)
    }

    static func previewRingtone(_ tone: Ringtone) {
        stopLoop()
        playOneShot(named: tone.rawValue, volume: 0.65)
    }
}

File: Android/app/src/main/java/net/mrblindbandit/app/connect/CallSounds.kt
package net.mrblindbandit.app.connect

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.SoundPool

/** UI / call SFX — full ambient ringtone + notification pack. */
class CallSounds(private val context: Context) {
    enum class Ringtone(val resName: String, val title: String) {
        DEFAULT("ringtone", "Default"),
        GOLD("ringtone_gold", "Gold"),
        PULSE("ringtone_pulse", "Pulse"),
        CHIME("ringtone_chime", "Chime"),
        AMBIENT_AURORA("ringtone_ambient_aurora", "Ambient · Aurora"),
        AMBIENT_BLOOM("ringtone_ambient_bloom", "Ambient · Bloom"),
        AMBIENT_GLASS("ringtone_ambient_glass", "Ambient · Glass"),
        AMBIENT_NIGHT("ringtone_ambient_night", "Ambient · Night"),
        AMBIENT_RAIN("ringtone_ambient_rain", "Ambient · Rain"),
        AMBIENT_SPACE("ringtone_ambient_space", "Ambient · Space")
    }

    enum class NotificationTone(val resName: String, val title: String) {
        MESSAGE("notif_message", "Message"),
        CALL("notif_call", "Call"),
        DELIVERED("notif_delivered", "Delivered"),
        READ("notif_read", "Read"),
        AMBIENT_SOFT("notif_ambient_soft", "Ambient · Soft"),
        AMBIENT_GLOW("notif_ambient_glow", "Ambient · Glow"),
        AMBIENT_DROP("notif_ambient_drop", "Ambient · Drop"),
        AMBIENT_PETAL("notif_ambient_petal", "Ambient · Petal"),
        AMBIENT_WIND("notif_ambient_wind", "Ambient · Wind")
    }

    var playUISounds: Boolean = true
    var selectedRingtone: Ringtone = Ringtone.DEFAULT
    var selectedNotificationTone: NotificationTone = NotificationTone.MESSAGE

    private val prefs = context.getSharedPreferences("blindbandit_sounds", Context.MODE_PRIVATE)
    private var pool: SoundPool? = null
    private val ids = mutableMapOf<String, Int>()
    private var ringbackPlayer: MediaPlayer? = null

    init {
        playUISounds = prefs.getBoolean("playUISounds", true)
        CallHaptics.enabled = prefs.getBoolean("hapticsEnabled", true)
        selectedRingtone = Ringtone.values().find { it.resName == prefs.getString("selectedRingtone", null) } ?: Ringtone.DEFAULT
        selectedNotificationTone = NotificationTone.values().find { it.resName == prefs.getString("selectedNotificationTone", null) }
            ?: NotificationTone.MESSAGE
        val attrs = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .build()
        pool = SoundPool.Builder().setMaxStreams(4).setAudioAttributes(attrs).build()
        val names = buildList {
            addAll(Ringtone.values().map { it.resName })
            addAll(NotificationTone.values().map { it.resName })
            addAll(
                listOf(
                    "call_initiated", "call_connected", "hangup", "busy_tone",
                    "message_sent", "voicenote_sent", "attachment_sent",
                    "typing_ping", "ringback_outgoing"
                )
            )
        }.distinct()
        names.forEach { name ->
            val resId = context.resources.getIdentifier(name, "raw", context.packageName)
            if (resId != 0) pool?.load(context, resId, 1)?.let { ids[name] = it }
        }
    }

    fun persist() {
        prefs.edit()
            .putBoolean("playUISounds", playUISounds)
            .putBoolean("hapticsEnabled", CallHaptics.enabled)
            .putString("selectedRingtone", selectedRingtone.resName)
            .putString("selectedNotificationTone", selectedNotificationTone.resName)
            .apply()
    }

    private fun play(name: String, volume: Float = 0.7f) {
        if (!playUISounds) return
        val id = ids[name] ?: return
        pool?.play(id, volume, volume, 1, 0, 1f)
    }

    fun startRingback() {
        if (!playUISounds) return
        stopRingback()
        val resId = context.resources.getIdentifier("ringback_outgoing", "raw", context.packageName)
        if (resId == 0) return
        ringbackPlayer = MediaPlayer.create(context, resId)?.apply {
            isLooping = true
            setVolume(0.42f, 0.42f)
            start()
        }
        CallHaptics.outgoingStart(context)
    }

    fun stopRingback() {
        try { ringbackPlayer?.stop() } catch (_: Exception) {}
        ringbackPlayer?.release()
        ringbackPlayer = null
    }

    fun callInitiated() { startRingback(); play("call_initiated", 0.55f) }
    fun callConnecting() = CallHaptics.connecting(context)
    fun callConnected() {
        stopRingback()
        play("call_connected", 0.55f)
        CallHaptics.connected(context)
    }
    fun hangup() {
        stopRingback()
        play("hangup", 0.65f)
        CallHaptics.disconnected(context)
    }
    fun busyOrFailed() {
        stopRingback()
        play("busy_tone", 0.55f)
        CallHaptics.busyOrFailed(context)
    }
    fun messageSent() { play("message_sent", 0.55f); CallHaptics.messageSent(context) }
    fun attachmentSent() { play("attachment_sent", 0.55f); CallHaptics.attachmentSent(context) }
    fun voiceNoteSent() { play("voicenote_sent", 0.6f); CallHaptics.voiceNoteSent(context) }
    fun typingPing() = play("typing_ping", 0.3f)
    fun inboundNotification() {
        play(selectedNotificationTone.resName, 0.6f)
        CallHaptics.inboundMessage(context)
    }
    fun delivered() { play("notif_delivered", 0.4f); CallHaptics.delivered(context) }
    fun read() { play("notif_read", 0.35f); CallHaptics.read(context) }
    fun previewRingtone(tone: Ringtone) = play(tone.resName, 0.65f)
    fun previewNotification(tone: NotificationTone) = play(tone.resName, 0.6f)
}
